src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -119,23 +125,103 @@ catch (const wil::ResultException& exception)
+        data.scale = _r.pixelPerDIP;
+        data.resolution.x = static_cast<float>(_r.cellCount.x * _r.fontMetrics.cellSize.x);
+        data.resolution.y = static_cast<float>(_r.cellCount.y * _r.fontMetrics.cellSize.y);
+        data.background = colorFromU32<f32x4>(_r.backgroundColor);
+
+#pragma warning(suppress : 26494) // Variable 'mapped' is uninitialized. Always initialize an object (type.5).
+        D3D11_MAPPED_SUBRESOURCE mapped;
+        THROW_IF_FAILED(_r.deviceContext->Map(_r.customShaderConstantBuffer.get(), 0, D3D11_MAP_WRITE_DISCARD, 0, &mapped));
+        assert(mapped.RowPitch >= sizeof(data));
+        memcpy(mapped.pData, &data, sizeof(data));
+        _r.deviceContext->Unmap(_r.customShaderConstantBuffer.get(), 0);
+    }
+
+    // Render with the custom shader.
+    {
+        // OM: Output Merger
+        // customOffscreenTextureView was just rendered to via customOffscreenTextureTargetView and is
+        // set as the output target. Before we can use it as an input we have to remove it as an output.
+        _r.deviceContext->OMSetRenderTargets(1, _r.renderTargetView.addressof(), nullptr);
+
+        // VS: Vertex Shader
+        _r.deviceContext->VSSetShader(_r.customVertexShader.get(), nullptr, 0);
+
+        // PS: Pixel Shader
+        _r.deviceContext->PSSetShader(_r.customPixelShader.get(), nullptr, 0);
+        _r.deviceContext->PSSetConstantBuffers(0, 1, _r.customShaderConstantBuffer.addressof());
+        _r.deviceContext->PSSetShaderResources(0, 1, _r.customOffscreenTextureView.addressof());
+        _r.deviceContext->PSSetSamplers(0, 1, _r.customShaderSamplerState.addressof());
+
+        _r.deviceContext->Draw(4, 0);
+    }
 
+    // For the next frame we need to restore our context state.
+    {
+        // VS: Vertex Shader
+        _r.deviceContext->VSSetShader(_r.vertexShader.get(), nullptr, 0);
+
+        // PS: Pixel Shader
+        _r.deviceContext->PSSetShader(_r.pixelShader.get(), nullptr, 0);
+        _r.deviceContext->PSSetConstantBuffers(0, 1, _r.constantBuffer.addressof());
+        const std::array resources{ _r.cellView.get(), _r.atlasView.get() };
+        _r.deviceContext->PSSetShaderResources(0, gsl::narrow_cast<UINT>(resources.size()), resources.data());
+        _r.deviceContext->PSSetSamplers(0, 0, nullptr);
+    }
+}
+
+void AtlasEngine::_setShaderResources() const
+{
+    // IA: Input Assembler
     // Our vertex shader uses a trick from Bill Bilodeau published in
     // "Vertex Shader Tricks" at GDC14 to draw a fullscreen triangle
     // without vertex/index buffers. This prepares our context for this.
     _r.deviceContext->IASetVertexBuffers(0, 0, nullptr, nullptr, nullptr);
     _r.deviceContext->IASetIndexBuffer(nullptr, DXGI_FORMAT_UNKNOWN, 0);
     _r.deviceContext->IASetInputLayout(nullptr);
-    _r.deviceContext->IASetPrimitiveTopology(D3D11_PRIMITIVE_TOPOLOGY_TRIANGLELIST);
+    _r.deviceContext->IASetPrimitiveTopology(D3D11_PRIMITIVE_TOPOLOGY_TRIANGLESTRIP);

```

Comment:
any meaningful difference here for Atlas w/o shaders?

Reply:
No, there shouldn't be any difference. Both a list and a strip require at least 3 vertices to define the initial rectangle. But specifying 4 vertices is invalid for a triangle list (which expects a list of distinct triangles) and valid for a strip (which appends the 4th and following vertex to the initial triangle to form a "strip").